File: tests/DesktopGrass.Win2D.Tests/SimTests/SheepGreetingTests.cs
// SheepGreetingTests.cs - §16 sheep proximity-greeting tests.
//
// Mirrors tests/DesktopGrass.Native.Tests/src/sheep_greeting_tests.cpp.

using System;
using System.Collections.Generic;
using System.Linq;
using DesktopGrass.Win2D;
using Xunit;

namespace DesktopGrass.Win2D.Tests.SimTests;

public class SheepGreetingTests
{
    private const double Monitor1920 = 1920.0;
    private const double EligibleAge = 2.0;
    private const double LongTimer = 10.0;

    private static Sim BuildSim(ulong seed = Constants.CANONICAL_TEST_SEED,
                                double monitorWidth = Monitor1920,
                                double density = 1.0)
    {
        var sim = new Sim
        {
            Blades = Sim.GenerateBlades(seed, monitorWidth, density),
            WindowHeight = Constants.STRIP_HEIGHT + Constants.HEADROOM,
            GroundY = Constants.STRIP_HEIGHT + Constants.HEADROOM,
        };
        sim.ResetAmbientGusts(seed, monitorWidth);
        sim.ResetEntities(seed);
        sim.SetCritter(CritterKind.Sheep);
        return sim;
    }

    private static List<int> SheepIndices(Sim sim)
    {
        var indices = new List<int>();
        for (int i = 0; i < sim.Entities.Count; i++)
        {
            if (sim.Entities[i].Kind == EntityKind.Sheep) indices.Add(i);
        }
        return indices;
    }

    private static void SetSheep(Sim sim, int index, double x, double vx,
                                 byte state = Constants.SHEEP_STATE_WALKING,
                                 double age = EligibleAge,
                                 double stateTimer = LongTimer)
    {
        Entity e = sim.Entities[index];
        e.X = x;
        e.Vx = vx;
        e.State = state;
        e.Age = age;
        e.StateTimer = stateTimer;
        sim.Entities[index] = e;
    }

    private static List<int> PrepareTwoSheep(Sim sim, double gap = 40.0,
                                             double ageA = EligibleAge,
                                             double ageB = EligibleAge)
    {
        List<int> indices = SheepIndices(sim);
        Assert.True(indices.Count >= 2);

        SetSheep(sim, indices[0], 500.0, -20.0, Constants.SHEEP_STATE_WALKING, ageA);
        SetSheep(sim, indices[1], 500.0 + gap, 18.0, Constants.SHEEP_STATE_WALKING, ageB);
        for (int n = 2; n < indices.Count; n++)
        {
            SetSheep(sim, indices[n], 1000.0 + 150.0 * n, 16.0);
        }
        return indices;
    }

    private static int AdvanceSidePastSheepGeneration(ref Prng side)
    {
        double countDraw = side.Uniform(Constants.SHEEP_COUNT_MIN, Constants.SHEEP_COUNT_MAX + 1);
        int expectedCount = (int)Math.Floor(countDraw);
        if (expectedCount < Constants.SHEEP_COUNT_MIN) expectedCount = Constants.SHEEP_COUNT_MIN;
        if (expectedCount > Constants.SHEEP_COUNT_MAX) expectedCount = Constants.SHEEP_COUNT_MAX;

        for (int i = 0; i < expectedCount; i++)
        {
            double margin = Constants.SHEEP_BODY_RADIUS + 8.0;
            _ = side.Uniform(margin, Monitor1920 - margin);
            _ = side.Uniform(Constants.SHEEP_WALK_SPEED_MIN, Constants.SHEEP_WALK_SPEED_MAX);
            _ = side.Uniform(0.0, 1.0);
            _ = side.NextU32();
            _ = side.Uniform(Constants.SHEEP_WALK_DURATION_MIN, Constants.SHEEP_WALK_DURATION_MAX);
        }
        return expectedCount;
    }

    [Fact]
    public void SheepGreetingConstantsArePinnedToSpecValues()
    {
        Assert.Equal((byte)5, Constants.SHEEP_STATE_GREETING);
        Assert.Equal(50.0, Constants.SHEEP_GREET_RADIUS);
        Assert.Equal(1.6, Constants.SHEEP_GREET_DURATION_MIN);
        Assert.Equal(2.8, Constants.SHEEP_GREET_DURATION_MAX);
        Assert.Equal(1.5, Constants.SHEEP_GREET_MIN_AGE);
        Assert.Equal(4.5, Constants.SHEEP_GREET_HEAD_BOB_FREQ);
        Assert.Equal(0.7, Constants.SHEEP_GREET_HEAD_BOB_AMP);
    }

    [Fact]
    public void SheepCuriousConstantsArePinnedToSpecValues()
    {
        Assert.Equal(80.0, Constants.SHEEP_CURIOUS_RADIUS);
        Assert.Equal(0.55, Constants.SHEEP_CURIOUS_HEAD_TURN_MAX);
    }

    [Theory]
    [InlineData(2, 0.70)]
    [InlineData(5, 0.70)]
    [InlineData(6, 0.10)]
    [InlineData(9, 0.10)]
    [InlineData(10, 0.30)]
    [InlineData(15, 0.30)]
    [InlineData(21, 0.30)]
    [InlineData(22, 0.70)]
    [InlineData(23, 0.70)]
    public void SheepSleepProbabilityFollowsLocalHourBands(int hour, double expected)
    {
        Assert.Equal(expected, Sim.SheepSleepProbForLocalHour(hour));
    }

    [Fact]
    public void EligibleNearbySheepEnterGreetingFacingEachOther()
    {
        Sim sim = BuildSim();
        List<int> indices = PrepareTwoSheep(sim);

        sim.TickEntities(0.016);

        Entity a = sim.Entities[indices[0]];
        Entity b = sim.Entities[indices[1]];
        Assert.Equal(Constants.SHEEP_STATE_GREETING, a.State);
        Assert.Equal(Constants.SHEEP_STATE_GREETING, b.State);
        Assert.InRange(a.StateTimer, Constants.SHEEP_GREET_DURATION_MIN, Constants.SHEEP_GREET_DURATION_MAX);
        Assert.Equal(a.StateTimer, b.StateTimer, 12);
        Assert.True(a.Vx > 0.0);
        Assert.True(b.Vx < 0.0);
    }

    [Fact]
    public void FarApartEligibleSheepDoNotGreet()
    {
        Sim sim = BuildSim();
        List<int> indices = PrepareTwoSheep(sim, 200.0);

        for (int i = 0; i < 3; i++) sim.TickEntities(0.016);

        Assert.Equal(Constants.SHEEP_STATE_WALKING, sim.Entities[indices[0]].State);
        Assert.Equal(Constants.SHEEP_STATE_WALKING, sim.Entities[indices[1]].State);
    }

    [Fact]
    public void SheepUnderGreetingMinimumAgeDoNotGreet()
    {
        Sim sim = BuildSim();
        List<int> indices = PrepareTwoSheep(sim, 40.0, 0.5, EligibleAge);

        sim.TickEntities(0.016);

        Assert.Equal(Constants.SHEEP_STATE_WALKING, sim.Entities[indices[0]].State);
        Assert.Equal(Constants.SHEEP_STATE_WALKING, sim.Entities[indices[1]].State);
    }

    [Fact]
    public void SleepingHoppingAndGreetingSheepAreNotGreetingEligible()
    {
        byte[] blockedStates =
        {
            Constants.SHEEP_STATE_SLEEPING,
            Constants.SHEEP_STATE_HOPPING,
            Constants.SHEEP_STATE_GREETING,
        };

        foreach (byte blockedState in blockedStates)
        {
            Sim sim = BuildSim();
            List<int> indices = PrepareTwoSheep(sim);
            SetSheep(sim, indices[0], 500.0, -20.0, blockedState, EligibleAge);

            sim.TickEntities(0.016);

            Assert.Equal(blockedState, sim.Entities[indices[0]].State);
            Assert.Equal(Constants.SHEEP_STATE_WALKING, sim.Entities[indices[1]].State);
        }
    }

    [Fact]
    public void GreetingExpiryReturnsSheepToWalkingWithVxFlipped()
    {
        Sim sim = BuildSim();
        List<int> indices = PrepareTwoSheep(sim);

        sim.TickEntities(0.016);
        double duration = sim.Entities[indices[0]].StateTimer;
        double aGreetingVx = sim.Entities[indices[0]].Vx;
        double bGreetingVx = sim.Entities[indices[1]].Vx;

        sim.TickEntities(duration + 0.01);

        Assert.Equal(Constants.SHEEP_STATE_WALKING, sim.Entities[indices[0]].State);
        Assert.Equal(Constants.SHEEP_STATE_WALKING, sim.Entities[indices[1]].State);
        Assert.Equal(-aGreetingVx, sim.Entities[indices[0]].Vx, 12);
        Assert.Equal(-bGreetingVx, sim.Entities[indices[1]].Vx, 12);
    }

    [Fact]
    public void GreetingTriggerConsumesOnePrngDrawPerPair()
    {
        var side = Prng.Init(Constants.CANONICAL_TEST_SEED ^ Constants.CRITTER_PRNG_SALT);

        Sim sim = BuildSim();
        int expectedCount = AdvanceSidePastSheepGeneration(ref side);
        Assert.Equal(expectedCount, SheepIndices(sim).Count);
        List<int> indices = PrepareTwoSheep(sim);

        double expectedDuration = side.Uniform(
            Constants.SHEEP_GREET_DURATION_MIN,
            Constants.SHEEP_GREET_DURATION_MAX);
        sim.TickEntities(0.016);

        Assert.Equal(expectedDuration, sim.Entities[indices[0]].StateTimer, 12);
        Assert.Equal(expectedDuration, sim.Entities[indices[1]].StateTimer, 12);
    }

    [Fact]
    public void SingleSheepCannotEnterGreeting()
    {
        Sim sim = BuildSim();
        Assert.True(sim.Entities.Count >= 1);
        sim.Entities.RemoveRange(1, sim.Entities.Count - 1);
        SetSheep(sim, 0, 500.0, 20.0);

        sim.TickEntities(0.016);

        Assert.Single(sim.Entities);
        Assert.Equal(Constants.SHEEP_STATE_WALKING, sim.Entities[0].State);
    }

    [Fact]
    public void ThreeSheepClusterGreetsOnlyTheFirstEncounteredPair()
    {
        Sim sim = BuildSim();
        List<int> indices = SheepIndices(sim);
        Assert.True(indices.Count >= 2);
        if (indices.Count < 3)
        {
            sim.Entities.Add(sim.Entities[indices[1]]);
            indices = SheepIndices(sim);
        }
        Assert.True(indices.Count >= 3);

        SetSheep(sim, indices[0], 500.0, -20.0);
        SetSheep(sim, indices[1], 540.0, 18.0);
        SetSheep(sim, indices[2], 580.0, 16.0);

        sim.TickEntities(0.016);

        Assert.Equal(Constants.SHEEP_STATE_GREETING, sim.Entities[indices[0]].State);
        Assert.Equal(Constants.SHEEP_STATE_GREETING, sim.Entities[indices[1]].State);
        Assert.Equal(Constants.SHEEP_STATE_WALKING, sim.Entities[indices[2]].State);
        Assert.Equal(2, sim.Entities.Count(e => e.Kind == EntityKind.Sheep
                                                && e.State == Constants.SHEEP_STATE_GREETING));
    }
}

File: tests/DesktopGrass.Native.Tests/src/sheep_greeting_tests.cpp
// sheep_greeting_tests.cpp
//
// §16 sheep proximity-greeting tests. Mirrors Win2D SheepGreetingTests.cs.

#include "TestHelpers.h"
#include "Sim.h"

#include <algorithm>
#include <cmath>
#include <cstddef>
#include <vector>

using namespace desktopgrass;

namespace {

constexpr double Monitor1920 = 1920.0;
constexpr double EligibleAge = 2.0;
constexpr double LongTimer = 10.0;

Sim build_sheep_sim() {
    Sim sim = sim_init(CANONICAL_TEST_SEED, Monitor1920, DEFAULT_DENSITY);
    sim_set_critter(sim, CritterKind::Sheep);
    return sim;
}

std::vector<std::size_t> sheep_indices(const Sim& sim) {
    std::vector<std::size_t> indices;
    for (std::size_t i = 0; i < sim.entities.size(); ++i) {
        if (sim.entities[i].kind == EntityKind::Sheep) indices.push_back(i);
    }
    return indices;
}

void set_sheep(Sim& sim, std::size_t index, double x, double vx,
               uint8_t state = SHEEP_STATE_WALKING,
               double age = EligibleAge,
               double stateTimer = LongTimer) {
    Entity& e = sim.entities[index];
    e.x = x;
    e.vx = vx;
    e.state = state;
    e.age = age;
    e.stateTimer = stateTimer;
}

std::vector<std::size_t> prepare_two_sheep(Sim& sim, double gap = 40.0,
                                           double ageA = EligibleAge,
                                           double ageB = EligibleAge) {
    std::vector<std::size_t> indices = sheep_indices(sim);
    Assert::IsTrue(indices.size() >= 2);

    set_sheep(sim, indices[0], 500.0, -20.0, SHEEP_STATE_WALKING, ageA);
    set_sheep(sim, indices[1], 500.0 + gap, 18.0, SHEEP_STATE_WALKING, ageB);
    for (std::size_t n = 2; n < indices.size(); ++n) {
        set_sheep(sim, indices[n], 1000.0 + 150.0 * static_cast<double>(n), 16.0);
    }
    return indices;
}

int advance_side_past_sheep_generation(Prng& side) {
    const double countDraw = prng_uniform(side, SHEEP_COUNT_MIN, SHEEP_COUNT_MAX + 1);
    int expectedCount = static_cast<int>(std::floor(countDraw));
    if (expectedCount < SHEEP_COUNT_MIN) expectedCount = SHEEP_COUNT_MIN;
    if (expectedCount > SHEEP_COUNT_MAX) expectedCount = SHEEP_COUNT_MAX;

    for (int i = 0; i < expectedCount; ++i) {
        const double margin = SHEEP_BODY_RADIUS + 8.0;
        (void)prng_uniform(side, margin, Monitor1920 - margin);
        (void)prng_uniform(side, SHEEP_WALK_SPEED_MIN, SHEEP_WALK_SPEED_MAX);
        (void)prng_uniform(side, 0.0, 1.0);
        (void)prng_next_u32(side);
        (void)prng_uniform(side, SHEEP_WALK_DURATION_MIN, SHEEP_WALK_DURATION_MAX);
        (void)prng_index(side, static_cast<uint32_t>(sizeof(SHEEP_NAME_POOL) / sizeof(SHEEP_NAME_POOL[0])));
    }
    return expectedCount;
}

int count_sheep_in_state(const Sim& sim, uint8_t state) {
    return static_cast<int>(std::count_if(sim.entities.begin(), sim.entities.end(),
        [state](const Entity& e) {
            return e.kind == EntityKind::Sheep && e.state == state;
        }));
}

} // namespace

using namespace Microsoft::VisualStudio::CppUnitTestFramework;

namespace DesktopGrassNativeTests
{
TEST_CLASS(SheepGreetingTests)
{
public:
TEST_METHOD(SheepGreetingConstantsArePinnedToSpecValues) {
    Assert::IsTrue(SHEEP_STATE_GREETING       == 5);
    Assert::IsTrue(SHEEP_GREET_RADIUS         == Near(50.0));
    Assert::IsTrue(SHEEP_GREET_DURATION_MIN   == Near(1.6));
    Assert::IsTrue(SHEEP_GREET_DURATION_MAX   == Near(2.8));
    Assert::IsTrue(SHEEP_GREET_MIN_AGE        == Near(1.5));
    Assert::IsTrue(SHEEP_GREET_HEAD_BOB_FREQ  == Near(4.5));
    Assert::IsTrue(SHEEP_GREET_HEAD_BOB_AMP   == Near(0.7));
}

TEST_METHOD(SheepCuriousConstantsArePinnedToSpecValues) {
    Assert::IsTrue(SHEEP_CURIOUS_RADIUS        == Near(80.0));
    Assert::IsTrue(SHEEP_CURIOUS_HEAD_TURN_MAX == Near(0.55));
}


TEST_METHOD(EligibleNearbySheepEnterGreetingFacingEachOther) {
    Sim sim = build_sheep_sim();
    const std::vector<std::size_t> indices = prepare_two_sheep(sim);

    sim_tick_entities(sim, 0.016);

    const Entity& a = sim.entities[indices[0]];
    const Entity& b = sim.entities[indices[1]];
    Assert::IsTrue(a.state == SHEEP_STATE_GREETING);
    Assert::IsTrue(b.state == SHEEP_STATE_GREETING);
    Assert::IsTrue(a.stateTimer >= SHEEP_GREET_DURATION_MIN);
    Assert::IsTrue(a.stateTimer <= SHEEP_GREET_DURATION_MAX);
    Assert::IsTrue(a.stateTimer == Near(b.stateTimer));
    Assert::IsTrue(a.vx > 0.0);
    Assert::IsTrue(b.vx < 0.0);
}

TEST_METHOD(FarApartEligibleSheepDoNotGreet) {
    Sim sim = build_sheep_sim();
    const std::vector<std::size_t> indices = prepare_two_sheep(sim, 200.0);

    for (int i = 0; i < 3; ++i) sim_tick_entities(sim, 0.016);

    Assert::IsTrue(sim.entities[indices[0]].state == SHEEP_STATE_WALKING);
    Assert::IsTrue(sim.entities[indices[1]].state == SHEEP_STATE_WALKING);
}

TEST_METHOD(SheepUnderGreetingMinimumAgeDoNotGreet) {
    Sim sim = build_sheep_sim();
    const std::vector<std::size_t> indices = prepare_two_sheep(sim, 40.0, 0.5, EligibleAge);

    sim_tick_entities(sim, 0.016);

    Assert::IsTrue(sim.entities[indices[0]].state == SHEEP_STATE_WALKING);
    Assert::IsTrue(sim.entities[indices[1]].state == SHEEP_STATE_WALKING);
}

TEST_METHOD(SleepingHoppingAndGreetingSheepAreNotGreetingEligible) {
    const uint8_t blockedStates[] = {
        SHEEP_STATE_SLEEPING,
        SHEEP_STATE_HOPPING,
        SHEEP_STATE_GREETING,
    };

    for (uint8_t blockedState : blockedStates) {
        Sim sim = build_sheep_sim();
        const std::vector<std::size_t> indices = prepare_two_sheep(sim);
        set_sheep(sim, indices[0], 500.0, -20.0, blockedState, EligibleAge);

        sim_tick_entities(sim, 0.016);

        Assert::IsTrue(sim.entities[indices[0]].state == blockedState);
        Assert::IsTrue(sim.entities[indices[1]].state == SHEEP_STATE_WALKING);
    }
}

TEST_METHOD(GreetingExpiryReturnsSheepToWalkingWithVxFlipped) {
    Sim sim = build_sheep_sim();
    const std::vector<std::size_t> indices = prepare_two_sheep(sim);

    sim_tick_entities(sim, 0.016);
    const double duration = sim.entities[indices[0]].stateTimer;
    const double aGreetingVx = sim.entities[indices[0]].vx;
    const double bGreetingVx = sim.entities[indices[1]].vx;

    sim_tick_entities(sim, duration + 0.01);

    Assert::IsTrue(sim.entities[indices[0]].state == SHEEP_STATE_WALKING);
    Assert::IsTrue(sim.entities[indices[1]].state == SHEEP_STATE_WALKING);
    Assert::IsTrue(sim.entities[indices[0]].vx == Near(-aGreetingVx));
    Assert::IsTrue(sim.entities[indices[1]].vx == Near(-bGreetingVx));
}

TEST_METHOD(GreetingTriggerConsumesOnePRNGDrawPerPair) {
    Prng side;
    prng_init(side, CANONICAL_TEST_SEED ^ CRITTER_PRNG_SALT);

    Sim sim = build_sheep_sim();
    const int expectedCount = advance_side_past_sheep_generation(side);
    Assert::IsTrue(static_cast<int>(sheep_indices(sim).size()) == expectedCount);
    const std::vector<std::size_t> indices = prepare_two_sheep(sim);

    const double expectedDuration = prng_uniform(side,
                                                 SHEEP_GREET_DURATION_MIN,
                                                 SHEEP_GREET_DURATION_MAX);
    sim_tick_entities(sim, 0.016);

    Assert::IsTrue(sim.entities[indices[0]].stateTimer == Near(expectedDuration));
    Assert::IsTrue(sim.entities[indices[1]].stateTimer == Near(expectedDuration));
}

TEST_METHOD(SingleSheepCannotEnterGreeting) {
    Sim sim = build_sheep_sim();
    sim.currentScene = Scene::Desert;
    Assert::IsTrue(sim.entities.size() >= 1);
    sim.entities.erase(sim.entities.begin() + 1, sim.entities.end());
    set_sheep(sim, 0, 500.0, 20.0);

    sim_tick_entities(sim, 0.016);

    Assert::IsTrue(sim.entities.size() == 1);
    Assert::IsTrue(sim.entities[0].state == SHEEP_STATE_WALKING);
}

TEST_METHOD(ThreeSheepClusterGreetsOnlyTheFirstEncounteredPair) {
    Sim sim = build_sheep_sim();
    std::vector<std::size_t> indices = sheep_indices(sim);
    Assert::IsTrue(indices.size() >= 2);
    if (indices.size() < 3) {
        sim.entities.push_back(sim.entities[indices[1]]);
        indices = sheep_indices(sim);
    }
    Assert::IsTrue(indices.size() >= 3);

    set_sheep(sim, indices[0], 500.0, -20.0);
    set_sheep(sim, indices[1], 540.0, 18.0);
    set_sheep(sim, indices[2], 580.0, 16.0);

    sim_tick_entities(sim, 0.016);

    Assert::IsTrue(sim.entities[indices[0]].state == SHEEP_STATE_GREETING);
    Assert::IsTrue(sim.entities[indices[1]].state == SHEEP_STATE_GREETING);
    Assert::IsTrue(sim.entities[indices[2]].state == SHEEP_STATE_WALKING);
    Assert::IsTrue(count_sheep_in_state(sim, SHEEP_STATE_GREETING) == 2);
}
};
}
